PPM-5 use client steering persistent data in optimal path

kantera800 opened the pull request:
This PR introduces client's steering restrictions handling in controller.

Client may have pre-configured settings stored in database such as:
"stay on initial radio"
"stay on selected band"

Optimal path task should check if current radio aligned with those configurations and manually steer to the correct band/radio if its required.

adam1985d commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2833,6 +2833,13 @@ bool db::clear_client_persistent_db(const sMacAddr &mac)
+    //todo need to change selected_bands type to eClientSelectedBands when merged to master
+    return (selected_bands ==
+            wireless_utils::which_freq(get_node_channel(tlvf::mac_to_string(hostap))));
+}
```
```suggestion
    //todo need to change selected_bands type to eClientSelectedBands when merged to master
    if (selected_bands == DISABLED || stay_on_seleceted_bands != eTriStateBool:Enabled) {
        return true;
    }
    return (selected_bands ==
            wireless_utils::which_freq(get_node_channel(tlvf::mac_to_string(hostap))));
}
```

itayx replied:
is `stay_on_seleceted_bands` global, or do we need to pass it along as an argument?

kantera800 replied:
aligned to `uint8_t selected_bands for now.
need to update eFreqType and use eClientSelectedBands format instead
https://jira.prplfoundation.org/browse/PPM-353

kantera800 committed:
optimal_path_task: remove duplicated hostap ssid validation

In case of non11k ssid validation is performed twice before
and after IRE candidates are added.
Remove the before part and validate only after candidate list is ready.

Signed-off-by: Alex Kanter <alex.kanter@devalore.com>

kantera800 committed:
optimal_path_task: add force steer flag

Preparative commit.
Sometimes clients repeatedly reject 11v steer requests.
Since we don't have any mechanism that will steer imminently after a
couple of tries we need the ability to do it manually.
Add a force steering flag.

Signed-off-by: Alex Kanter <alex.kanter@devalore.com>

kantera800 committed:
db: add method is_hostap_on_selected_bands

Preparative commit.
Checks if the radio's band is on one of the selected bands.

Signed-off-by: Alex Kanter <alex.kanter@devalore.com>

kantera800 committed:
optimal_path_task: handle client's band/radio steering restrictions

Client's steering restrictions may be stored in the database.
Add optimal path support for "stay on initial radio" and
"stay on selected band" parameters.
Check that current radio aligned with those parameters or
manually steer to the correct band/radio.

Signed-off-by: Alex Kanter <alex.kanter@devalore.com>

kantera800 committed:
association handling task: increase rssi first measurment request delay

When client is associated for the first time sometimes it takes
more than  5sec for it to appear in the ARP table.
ARP table is used to get client's IP (in case there was no
 DHCP notification handling).
In order to measure rssi we send ARP requests those will fail without
a valid destination IP address.
Increase Initial rssi measurement delay to 10sec.

Signed-off-by: Alex Kanter <alex.kanter@devalore.com>

itayx commented on controller/src/beerocks/master/tasks/optimal_path_task.cpp:
```diff
@@ -227,8 +227,86 @@ void optimal_path_task::work()
+                                         selected_bands, tlvf::mac_from_string(sibling));
+                                 });
+
+                if (it == current_hostap_siblings.end()) {
```
Shouldn't the check be:
```suggestion
                if (it != current_hostap_siblings.end()) {
```
Otherwise, it will set the `chosen_bssid` if there is no match

kantera800 replied:
fixed

morantr commented on controller/src/beerocks/master/tasks/optimal_path_task.cpp:
```diff
@@ -1207,8 +1207,7 @@ void optimal_path_task::work()
 
```
Commit message has typos:
"mechanizm" -> "mechanism"
"tryies" -> "tries"
"manually" -> "manually"
"Adding force steer flag" -> "Add a force steer flag"

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2737,6 +2737,21 @@ bool db::clear_client_persistent_db(const sMacAddr &mac)
+        return false;
+    }
+}
+
```
Why did you put this function on the database? It is not related to the database since it is not accessing any database private function. This function is used only by optimal path task, and therefore should be located there.

kantera800 replied:
i does access get_node_channel() .
but now that i think of i should probably change the signature to 
`bool db::is_hostap_on_client_selected_bands(const sMacAddr client, const sMacAddr &hostap)`

morantr replied:
So what it access to `get_node_channel()`? It is a public database function and can be accessed from outside the database.
You can do a global search and see that it used on many places from outside the database.
just instead of call it directly, you would have to add `database.` before: `database.get_node_channel()`.
I really believe that the database is not the right place for this function and that.

kantera800 replied:
i changed the signature to  ` db::is_hostap_on_client_selected_bands(const sMacAddr client, const sMacAddr &hostap)`
since this function can be uses from other tasks it belongs to the db (for example client_association_task )

morantr commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -766,6 +766,14 @@ class db {
+     * @param hostap MAC address of a radio.
+     * @return true on success, otherwise false.
+     */
+    bool is_hostap_on_client_selected_bands(const int8_t selected_bands, const sMacAddr &hostap);
```
Why using `int8_t` for something which is enum? It will be more clear if it will be an enum type on the function signature.

kantera800 replied:
cant use bitmap with an enum

morantr replied:
You can. Enum can receive any value which in range between its min to max value.
https://wiki.sei.cmu.edu/confluence/display/cplusplus/INT50-CPP.+Do+not+cast+to+an+out-of-range+enumeration+value

kantera800 replied:
you still need some kind of  `intVar` to pass the data, here we use `selected_bands` 
there is no getting around this.
also can be refactored in https://jira.prplfoundation.org/browse/PPM-353

morantr commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -766,6 +766,14 @@ class db {
+     * @param hostap MAC address of a radio.
+     * @return true on success, otherwise false.
+     */
+    bool is_hostap_on_client_selected_bands(const int8_t selected_bands, const sMacAddr &hostap);
```
In general, we try to remove some keywords from the code and replace them with more generic keywords.
`hostap` is one of them, and it needs to be replaced with `radio`.
I would change the function name to: `is_radio_band_included_on_client_selected_bands`. It is little bit longer but more clear.

kantera800 replied:
I agree , but then you whould see somthing like `is_radio_on_client_selected_bands(selected_bands, **hostap**)` in the code and that is confusing .
Anyway I always rename `hostap`  whenever its doesn't refer to a radio (in cases like VAPs IREs etc..)

morantr commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -766,6 +766,14 @@ class db {
 
+    /**
+     * @brief Check if the radio's band is on one of the selected bands.
+     * @param selected_bands Client selected band/bands. Possible values are bitwise options of eClientSelectedBands.
```
Missing space between the `@brief` and `@param`.

morantr replied:
This comment "...bitwise options of eClientSelectedBands." exceeding 100 characters ruler.
You can add a ruler to VSCODE, that will show where is the 100th column.

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2737,6 +2737,21 @@ bool db::clear_client_persistent_db(const sMacAddr &mac)
+    case beerocks::eFreqType::FREQ_5G:
+        return (selected_bands & eClientSelectedBands::eSelectedBands_5G);
+    default:
+        LOG(WARNING) << "hostap band " << hostap_band << "  is not supported by client";
```
```suggestion
        LOG(WARNING) << "hostap band " << hostap_band << " is not supported by client";
```

morantr commented on controller/src/beerocks/master/tasks/optimal_path_task.cpp:
```diff
@@ -227,8 +227,88 @@ void optimal_path_task::work()
                 TASK_LOG(INFO) << "Remove candidate " << it->first << " , vap " << candidate_bssid
```
There are some typos on the commit message and excessive spaces:

"**A** client's steering restrictions may be stored in **the** database.
Add optimal path support for "stay on initial radio" and
"stay on selected band" parameters.
Check that current radio aligned with those parameters or
**manually** steer to the correct band/radio."

morantr commented on controller/src/beerocks/master/tasks/optimal_path_task.cpp:
```diff
@@ -227,8 +227,88 @@ void optimal_path_task::work()
-            ++it;
+        }
+
+        // Check client's steering persistant database for any band/radio/device restrictions
```
```suggestion
        // Check client's steering persistent database for any band/radio/device restrictions.
```

morantr commented on controller/src/beerocks/master/tasks/optimal_path_task.cpp:
```diff
@@ -227,8 +227,88 @@ void optimal_path_task::work()
+                finish();
+                return;
+            }
+            // Client is not on initial radio, lets try to find it in the steering potential candidate list
```
```suggestion
            // Client is not on initial radio. Try to find it on the steering candidates list.
```

morantr commented on controller/src/beerocks/master/tasks/optimal_path_task.cpp:
```diff
@@ -227,8 +227,88 @@ void optimal_path_task::work()
+        auto client = tlvf::mac_from_string(sta_mac);
+        if (database.get_client_stay_on_initial_radio(client) == eTriStateBool::ENABLE) {
+            TASK_LOG(INFO) << "Client stay on initial radio enabled";
+            auto client_initial_hostap = database.get_client_initial_radio(client);
```
Please don't use `hostap`. Replace it to `radio`.

morantr commented on controller/src/beerocks/master/tasks/optimal_path_task.cpp:
```diff
@@ -227,8 +227,88 @@ void optimal_path_task::work()
+                              << " is not on the candidate ap list, continue as usual.";
+        }
+
+        if (database.get_client_stay_on_selected_band(client) == eTriStateBool::ENABLE) {
```
On line 237 you ask exactly the same condition and end it just before this line, so it seems to me that you can remove this line and the closing parenthesis of the last condition.

kantera800 replied:
different function get_client_stay_on_**initial_radio**()

morantr commented on controller/src/beerocks/master/tasks/optimal_path_task.cpp:
```diff
@@ -227,8 +227,88 @@ void optimal_path_task::work()
             }
-            ++it;
+        }
+
```
One of the reasons optimal path code is so hard to maintain is because it has such long functions.
This function is already big, I think it will be more correct to create a function for the cod you've added, or at least a big paragraph explaining what are you doing here and why.

kantera800 replied:
believe me i tried, but there has to be  3 possible outcomes continue/break and end task 
so single function solution will not look pretty.
Added to optimal path refactoring task https://jira.prplfoundation.org/browse/PPM-102

morantr commented on controller/src/beerocks/master/tasks/optimal_path_task.cpp:
```diff
@@ -923,6 +1004,92 @@ void optimal_path_task::work()
+        // Check client's steering persistant database for any band/radio/device restrictions
+        // This code is duplicated for both FIND_AND_PICK_HOSTAP_CROSS FIND_AND_PICK_HOSTAP_11K
+        // TODO: need to create preliminary state that prepares candidate hostap list
+        // regardless of client's 11k support  https://jira.prplfoundation.org/browse/PPM-102
```
```suggestion
        // Check client's steering persistent database for any band/radio/device restrictions
        // This code is duplicated for both FIND_AND_PICK_HOSTAP_CROSS FIND_AND_PICK_HOSTAP_11K
        // TODO: Need to create a preliminary state that prepares candidate radio list
        // regardless of client's 11k support #PPM-102.
```

morantr commented on controller/src/beerocks/master/tasks/optimal_path_task.cpp:
```diff
@@ -923,6 +1004,92 @@ void optimal_path_task::work()
+                finish();
+                return;
+            }
+            // Client is not on initial radio, lets try to find it in the steering potential candidate list
```
```suggestion
            // Client is not on initial radio. Try to find it in the steering candidates list.
```
The fact that you did copy-paste of the code above, is enforcing the requirement to add this to a function.

kantera800 replied:
see https://github.com/prplfoundation/prplMesh/pull/1560#discussion_r465831111

morantr commented on controller/src/beerocks/master/tasks/optimal_path_task.cpp:
```diff
@@ -923,6 +1004,92 @@ void optimal_path_task::work()
+                              << " is not on the candidate ap list, continue as usual.";
+        }
+
+        if (database.get_client_stay_on_selected_band(client) == eTriStateBool::ENABLE) {
```
Same condition as on 1024.

kantera800 replied:
nope

morantr commented on controller/src/beerocks/master/tasks/optimal_path_task.cpp:
```diff
@@ -989,17 +965,26 @@ void optimal_path_task::work()
         }
 
-        // Check if hostapd has suitable ssid
+        // Check if hostap has suitable ssid
```
```suggestion
        // Check if the radio has suitable SSID.
```

kantera800 replied:
again hostapd -> radio renaming should be done in separate PR

morantr commented:
All commit messages missing reference to Jira PPM-5.

adam1985d commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2737,6 +2737,22 @@ bool db::clear_client_persistent_db(const sMacAddr &mac)
+bool db::is_hostap_on_client_selected_bands(const sMacAddr &client, const sMacAddr &hostap)
+{
+    auto hostap_band    = wireless_utils::which_freq(get_node_channel(tlvf::mac_to_string(hostap)));
+    auto selected_bands = get_client_selected_bands(client);
+
```
```suggestion
    auto selected_bands = get_client_selected_bands(client);
    if (selected_bands == eClientSelectedBands::DISABLED) {
        return true;
    }

```

kantera800 replied:
will be handled in https://jira.prplfoundation.org/browse/PPM-353

adam1985d commented on controller/src/beerocks/master/tasks/optimal_path_task.cpp:
```diff
@@ -227,8 +227,88 @@ void optimal_path_task::work()
+                return;
+            }
+            // Client is not on initial radio, lets try to find it in the steering potential candidate list
+            if (potential_11k_aps.find(tlvf::mac_to_string(client_initial_radio)) !=
+                potential_11k_aps.end()) {
+                // Initial client hostap is on the candidate list, force steer the client there.
+                chosen_bssid = database.get_hostap_vap_with_ssid(
+                    tlvf::mac_to_string(client_initial_radio), current_hostap_ssid);
+                state          = SEND_STEER_ACTION;
+                is_force_steer = true;
+                TASK_LOG(INFO) << "Steer client imminently to initial radio "
+                               << client_initial_radio;
+                break;
+            }
+            TASK_LOG(WARNING) << "Client's initial radio " << client_initial_radio
+                              << " is not on the candidate ap list, continue as usual.";
```
in general - there are many places you use `return;` and other places you use `break;` be consistent about it

adam1985d commented on controller/src/beerocks/master/tasks/optimal_path_task.cpp:
```diff
@@ -227,8 +227,88 @@ void optimal_path_task::work()
+        }
+
+        if ((hostap_candidates_size == 1) && (potential_11k_aps.begin()->first == current_hostap)) {
+            TASK_LOG(WARNING)
```
```suggestion
            TASK_LOG(DEBUG)
```

kantera800 replied:
why?

kantera800 replied:
never mind , changed to debug

adam1985d replied:
since WARNING should indicate that this situation should not be reached, but we can waver and not fail on it.
in this case, I think the scenario is a feasible scenario - if you think it doesn't leave it as WARNING.

adam1985d commented on controller/src/beerocks/master/tasks/optimal_path_task.cpp:
```diff
@@ -1756,3 +1923,29 @@ bool optimal_path_task::is_hostap_on_cs_process(const std::string &hostap_mac)
+        TASK_LOG(ERROR) << "Candidate list is empty, nothing to remove";
+        return;
+    }
+    TASK_LOG(INFO) << "Stay on selected band enabled";
```
not required here - please remove

adam1985d commented on controller/src/beerocks/master/tasks/optimal_path_task.cpp:
```diff
@@ -989,17 +965,26 @@ void optimal_path_task::work()
+            }
+
+            // Steering allowed on all vaps unless load_steer_on_vaps list is defined
+            // on the platform , in that case verify that vap is on that list
```
```suggestion
            // on the platform, in that case, verify that vap is on that list
```

adam1985d commented on controller/src/beerocks/master/tasks/optimal_path_task.cpp:
```diff
@@ -989,17 +965,26 @@ void optimal_path_task::work()
-                               << current_hostap_ssid;
+            std::string candidate_bssid =
+                database.get_hostap_vap_with_ssid(it->first, current_hostap_ssid);
+
```
line-space can be removed
(nitpick)

adam1985d reviewed (approved):
small fixes - no re-review is required so approving.

morantr reviewed (approved):
Approving, but I unresolved to issues:
https://github.com/prplfoundation/prplMesh/pull/1560#discussion_r467712383
https://github.com/prplfoundation/prplMesh/pull/1560#discussion_r467715760

So please consider them.

kantera800 commented:
> All commit messages missing reference to Jira [PPM-5](https://jira.prplfoundation.org/browse/PPM-5).

you don't have to add  JIRA # in the commit title

kantera800 commented:
> Approving, but I unresolved to issues:
> [#1560 (comment)](https://github.com/prplfoundation/prplMesh/pull/1560#discussion_r467712383)
> [#1560 (comment)](https://github.com/prplfoundation/prplMesh/pull/1560#discussion_r467715760)
> 
> So please consider them.

replied
